### PR TITLE
Explore: Fix URL sync with async queries import 

### LINKED ISSUE #78115: Explore: datasources in the URL out of sync
Reproduction steps. **You must follow these steps exactly by navigating initially with the link.** Navigating back and forth between loki and slow-prometheus does not consistently reproduce the issue. Navigating from the loki link to normal prometheus does not reproduce the issue.

1. Run devenv with normal loki and slow proxy prometheus. I used `make devenv sources=tempo,slow_proxy`
2. Go to this [link](http://localhost:3000/explore?panes=%7B%22yC0%22:%7B%22datasource%22:%22gdev-loki%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bcompose_service%3D%5C%22db%5C%22%7D%20%7C%3D%20%60%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22gdev-loki%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%22now-6h%22,%22to%22:%22now%22%7D%7D%7D&schemaVersion=1&orgId=1) 
3. Wait for the loki query to finish
4. Change the datasource to slow-prometheus
5. Slow prometheus' query editor will show with the loki options (IE with the label browser). The query header will also say "(gdev-loki)"

--- 

This occurs because the url sync logic fires multiple times in this scenario. The history after the above steps are taken looks like this 

1. Pane DS loki, query DS loki : `explore?panes={"lR_":{"datasource":"loki-ds","queries":[{"refId":"A","expr":"{app=\"aggregator\"} | logfmt","queryType":"range","datasource":{"type":"loki","uid":"loki-ds"},"editorMode":"code"}],"range":{"from":"now-6h","to":"now"}}}&schemaVersion=1&orgId=1`
2. Pane DS prom, query DS prom : `explore?panes={"lR_":{"datasource":"prom-ds","queries":[{"refId":"A","expr":"{app=\"aggregator\"}","range":true,"datasource":{"type":"prometheus","uid":"prom-ds"}}],"range":{"from":"now-6h","to":"now"}}}&schemaVersion=1&orgId=1`
3.  Pane DS prom, query DS loki : `explore?panes={"lR_":{"datasource":"prom-ds","queries":[{"refId":"A","expr":"{app=\"aggregator\"} | logfmt","queryType":"range","datasource":{"type":"loki","uid":"loki-ds"},"editorMode":"code"}],"range":{"from":"now-6h","to":"now"}}}&schemaVersion=1&orgId=1`

Under functional navigations, the previous params and current params are never mismatched, so the sync doesn't occur. With the above, the URL appears to be out of sync, so the sync logic fires and causes the issue.

### PR DESCRIPTION
**Pull request by Elfo404**
**What is this feature?**

Fixes URL sync in explore when changing datasources causes long running import operations.

While this technically work, it surfaces other issues. now we cancel queries when changing ds (or navigating via the browser history) however some datasources perform network operationa s part of their initialization process, and there's currently no way to stop them. this results in browsers reaching the max number of request against a single host and - given we need to make an additional request to load correlations when switching datasource - the UI is blocked until all of them finish.

TLDR; we need a way to cancel EVERY pending request originating from datasources, which is currently missing.
It is unclear if we can use `backendSrv` to handle this transparently from datasource or if we need to some other teardown api.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #78115

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/explore/hooks/useStateSync/index.ts [hunk 1]
```diff
@@ -47,20 +48,40 @@ export function useStateSync(params: ExploreQueryParams) {
...
+          has been dispatched and pausing the listener until the changeDatasource.fulfilled action is dispatched.
+        */
+
+          {
+            if (changeDatasource.pending.type === action.type) {
+              paused.current = true;
+            }
+            if (changeDatasource.fulfilled.type === action.type) {
+              paused.current = false;
+            }
```
Comment (Elfo404) [inline public/app/features/explore/hooks/useStateSync/index.ts:70]:
```suggestion
            paused.current = changeDatasource.pending.type === action.type;
```

File: public/app/features/explore/hooks/useStateSync/index.ts [hunk 2]
```diff
@@ -47,20 +48,40 @@ export function useStateSync(params: ExploreQueryParams) {
...
-            changeRangeAction.type,
-            changePanelsStateAction.type,
-          ].includes(action.type),
+          /*
+          We want to update the URL when:
+           - a pane is opened or closed
+           - a query is run
+           - range is changed
+           - panel state is updated
+           - a dtasource change has completed.
```
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync/index.ts:57]:
```suggestion
           - a datasource change has completed.
```

**Review by ifrost — COMMENTED** [review]
Seems to work as expected 👍  Let's create a follow up ticket to improve tearing down async actions while data source changes. Regarding the test: maybe we could have a test that dispatches `changeDatasource.pending` and check if any actions dispatched after that do not change the URL until `changeDatasource.fullfied` is dispatched? A bit whiteboxing but at least should cover that we have pausing logic in place in case we refactor this component in future.

**Review by ifrost — APPROVED** [review]
